src/authn/ldap.js

```diff
@@ -0,0 +1,74 @@
+import express from 'express'
+import local from 'passport-local'
+import passport from 'passport'
+import assert from 'assert'
+import User from './../user'
+import LDAPClient from './../ldap';
+
+class LDAPLogin {
+  constructor(options) {
+    assert(options, 'options are required');
+    assert(options.cfg, 'options.cfg is required');
+    assert(options.cfg.ldap, 'options.cfg.ldap is required');
+    assert(options.cfg.ldap.url, 'options.cfg.ldap.url is required');
+    assert(options.cfg.ldap.cert, 'options.cfg.ldap.cert is required');
+    assert(options.cfg.ldap.key, 'options.cfg.ldap.key is required');
+    assert(options.authorize, 'options.authorize is required');
+
+    this.user = options.cfg.ldap.user;
+    this.password = options.cfg.ldap.password;
+    this.client = new LDAPClient(options.cfg.ldap);
+    this.authorize = options.authorize;
+
+    passport.use(new local.Strategy({
+      passReqToCallback: true,
+    }, (req, user, password, done) => {
+          this.localCallback(req, user, password)
+          .then((res) => done(null, res))
+          .catch((err) => {
+            console.log("Error authenticating local user", err.stack);
+            done(null, false, {message: err.message});
+          })
+        }));
+  }
+
+  router() {
+    let router = new express.Router();
+    router.post('/login', passport.authenticate('local', {
+      successRedirect: '/',
+      failureRedirect: '/',
+      failureFlash: true
+    }));
+    return router;
+  };
+
+  authFail() {
+    // we do not want to distinguish no-such-user from bad-password, lest this
+    // tool become an oracle for discovering valid usernames
+    throw new Error("User not found or incorrect password");
+  }
+
+  async localCallback(req, email, password) {
+    // bind as the bind user to convert email to DN
+    let userDn = await this.client.bind(this.user, this.password,
+        (client) => client.dnForEmail(email));
+    if (!userDn) {
+      return this.authFail();
+    }
+
+    // re-bind as the DN, to validate the password
+    try {
+      await this.client.bind(userDn, password);
```

Comment:
Generational, copying garbage collectors, and other tricks makes this all too fun... Coupled with immutable strings, I think it's safe to argue this would be hard.

We have to trust the node vm does the right thing (tm). We have no choice...
Maybe there is a way, probably if we used ArrayBuffers, but we probably can't because they library takes strings, and rewriting it... all the way down to where it is passed to the net driver is unrealistic... Even then node could still make a copy somewhere we don't know.

On the bright side, heroku is responsible for keeping this isolated. And the machine isn't directly exposed to internet, all data goes through a reverse proxy... So attacks are hard...